src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp

```diff
@@ -336,41 +387,54 @@ class boss_archimonde : public CreatureScript
+        }
 
         bool CanUseFingerOfDeath()
         {
-            // First we check if our current victim is in melee range or not.
-            Unit* victim = me->GetVictim();
-            if (victim && me->IsWithinDistInMap(victim, me->GetAggroRange(victim)))
-                return false;
-
-            ThreatContainer::StorageType const &threatlist = me->getThreatManager().getThreatList();
-            if (threatlist.empty())
-                return false;
-
-            std::list<Unit*> targets;
-            ThreatContainer::StorageType::const_iterator itr = threatlist.begin();
-            for (; itr != threatlist.end(); ++itr)
+            if (me->IsAlive())
             {
-                Unit* unit = ObjectAccessor::GetUnit(*me, (*itr)->getUnitGuid());
-                if (unit && unit->IsAlive())
-                    targets.push_back(unit);
-            }
+                // First we check if our current victim is in melee range or not.
+                Unit* victim = me->GetVictim();
+                if (victim && me->IsWithinDistInMap(victim, me->GetAggroRange(victim)))
+                    return false;
 
-            if (targets.empty())
-                return false;
+                ThreatContainer::StorageType const &threatlist = me->getThreatManager().getThreatList();
+                if (threatlist.empty())
+                    return false;
 
-            targets.sort(Trinity::ObjectDistanceOrderPred(me));
-            Unit* target = targets.front();
-            if (target)
-            {
-                if (!me->IsWithinDistInMap(target, me->GetAggroRange(target)))
-                    return true;                                // Cast Finger of Death
-                else                                            // This target is closest, he is our new tank
-                    me->AddThreat(target, me->getThreatManager().getThreat(me->GetVictim()));
-            }
+                ThreatContainer::StorageType::const_iterator itr = threatlist.begin();
+                for (; itr != threatlist.end(); ++itr)
+                {
+                    Unit* unit = ObjectAccessor::GetUnit(*me, (*itr)->getUnitGuid());
+                    if (unit && unit->IsAlive() && unit->IsCharmedOwnedByPlayerOrPlayer())
+                        targets.push_back(unit);
+                }
 
+                if (targets.empty())
+                    return false;
+
+                targets.sort(Trinity::ObjectDistanceOrderPred(me));
+                Unit* target = targets.front();
+                if (target)
+                {
+                    if (!me->IsWithinDistInMap(target, me->GetAggroRange(target)))
+                        return true;                                // Cast Finger of Death
+                    else                                            // This target is closest, he is our new tank
+                        me->AddThreat(target, me->getThreatManager().getThreat(me->GetVictim()));
```

Comment:
Should I remove the else?

Reply:
Yes, the `AddThreat()` line at least. If you do the above review, you could easily just remove this whole closure.

Reply:
If this method is just to check if Archimonde *can cast Finger of Death* you shouldn't be handling the threat list here.